python/ffsim/states/sample_slater.py

```diff
@@ -211,3 +403,96 @@ def _generate_probs(rdm: np.ndarray, sample: set[int], norb: int) -> np.ndarray:
         indices = list(sample | {orb})
         probs[i] = np.linalg.det(rdm[indices][:, indices]).real
     return probs / np.sum(probs)
+
+
+class _FastProjectionSampler:
+    """Projection-based fast sampler core with cached precomputation.
+
+    Given an orthonormal orbital matrix for a projector 1-RDM, the sampler
+    precomputes the normal vectors needed for the Gaussian-elimination sampling
+    routine and exposes a ``sample`` method that generates integer-encoded
+    bitstrings using a provided RNG.
+    """
+
+    def __init__(self, orbitals: np.ndarray):
+        """Initialize the sampler with occupied orbitals from a projector 1-RDM.
+
+        Precomputes dual normal vectors via reduced QR factorization and
+        triangular solves for reuse during sampling.
+
+        Args:
+            orbitals: A 2D array whose columns are orthonormal orbitals spanning
+                the occupied subspace. Shape ``(norb, nelec)``.
+
+        Raises:
+            ValueError: If the input is not 2D, has zero occupied orbitals, or is
+                ill-conditioned for the QR-based precomputation.
+        """
+        if orbitals.ndim != 2:
+            raise ValueError("Projector orbitals must be a 2D array.")
+        self._norb, self._nelec = orbitals.shape
+        if self._nelec == 0:
+            raise ValueError("Fast sampler requires at least one occupied orbital.")
+        # Precompute normal vectors h_t from QR
+        q, r = np.linalg.qr(orbitals, mode="reduced")
+        diag = np.abs(np.diag(r))
+        tol = 1e-12
```

Comment:
Generally, tolerances like this should be passed as an argument to the function, with a default value if appropriate.